File: Day181_2025/Q181_GFG.py
class Solution:
    def canReachHeight(self, arr, k, w, target):
        n = len(arr)
        water = [0] * (n + 1)
        ops = 0
        curr = 0

        for i in range(n):
            curr += water[i]
            effective_height = arr[i] + curr

            if effective_height < target:
                add = target - effective_height
                ops += add
                if ops > k:
                    return False

                curr += add
                if i + w < n:
                    water[i + w] -= add

        return True

    def maxMinHeight(self, arr, k, w):
        low = min(arr)
        high = low + k
        answer = low

        while low <= high:
            mid = (low + high) // 2
            if self.canReachHeight(arr, k, w, mid):
                answer = mid
                low = mid + 1
            else:
                high = mid - 1

        return answer

File: Day181_2025/Q181_GFG.cpp
class Solution {
    bool canReachHeight(vector<int>& arr, int k, int w, int target) {
        int n = arr.size();
        vector<int> water(n + 1, 0);
        long long ops = 0;
        long long curr = 0;

        for (int i = 0; i < n; ++i) {
            curr += water[i];
            int effectiveHeight = arr[i] + curr;

            if (effectiveHeight < target) {
                int add = target - effectiveHeight;
                ops += add;
                if (ops > k) return false;

                curr += add;
                if (i + w < n)
                    water[i + w] -= add;
            }
        }
        return true;
    }
  public:
    int maxMinHeight(vector<int> &arr, int k, int w) {
        // Q. Max min Height
        // code here
        int mn = *min_element(arr.begin(), arr.end());
            int low = *min_element(arr.begin(), arr.end());
        int high = low + k;
        int answer = low;

        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (canReachHeight(arr, k, w, mid)) {
                answer = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return answer;
    }
};